### DISCUSSION TITLE
Full Framework 4.7 & Core 2.1 project in solution, Frosting the only runner I can use?

### DISCUSSION THREAD

**Discussion by klabranche** [cake-build/cake#3002] [Q&A] [answered]
I'm currently on cake 0.33, was looking to go to 1.0.

Am I reading the docs right?  I'd have to move to cake frosting to get a single build script as I current have atm in 0.33?

**Comment by pascalberger** [comment, marked as answer]
There hasn't changed anything in this regard between 0.33 and 1.0. 

There are [different runners](https://cakebuild.net/docs/running-builds/runners/) available for Cake. [Cake.Frosting](https://cakebuild.net/docs/running-builds/runners/cake-frosting), which allows to write your build script as console applications is one of those. But you still can use any of the script runners, like [Cake .NET Tool](https://cakebuild.net/docs/running-builds/runners/dotnet-tool) or [Cake Runner for .NET Framework](https://cakebuild.net/docs/running-builds/runners/cake-runner-for-dotnet-framework). [Cake Runner for .NET Core](https://cakebuild.net/docs/running-builds/runners/cake-runner-for-dotnet-core) is also still available if you need to run on .NET Core 2.0. For newer .NET Core versions [Cake .NET Tool](https://cakebuild.net/docs/running-builds/runners/dotnet-tool) is the recommended way to run Cake scripts. These runners were also already available like this in 0.33. See [Cake Runners](https://cakebuild.net/docs/running-builds/runners/) for an overview of the available runners and supported platforms.

**Reply by klabranche** [reply]
So I could just change my pin from .33 to 1.0 (RC) and should be good to go (in general, understand addins, etc... may not be updated, etc...)?  And understand the changes as outlined in https://cakebuild.net/docs/getting-started/upgrade...

**Reply by pascalberger** [reply]
Yes. If you use obsolete members you need to update your script and there are some breaking changes in the CLI outlined in the document you mentioned. If you use one of the script runners the Cake.Frosting section won't affect you.

**Reply by klabranche** [reply]
Understood.  Thanks for the clarity.  It didn't appear from the documentation that I had a path to go from .3X ("traditional/original" runner is the way I would put it) to 1.0.  Just upped to 1.0 and just had to update to using --configuration & --target instead of -c -t.  

Love Cake.  Thanks for all you and your team do. :-)

**Reply by pascalberger** [reply]
Which part of the documentation was not clear to you regarding supported runners? 

Script hosts and Console application host (Frosting) are two different approaches for writing build scripts, each with its own pros and cons. We don't consider script hosts outdated. Our recommended way to run Cake scripts is Cake .NET Tool, which is a script runner, since this is what seems to fit best for most people. But Cake is all about choice and you should pick the approach which fits the best in your specific scenario.

**Reply by klabranche** [reply]
Understood, on difference with script hosts and console app (frosting).

When I looked at the runners and supported platforms specifically the supported platforms table (https://cakebuild.net/docs/running-builds/runners/) it appears that Cake Frosting is the only one that could do a Full Framework project AND Core project (as I have 1 solution with 4 FF projects, and 2 Core projects (2.1)).  

For me at least, I wasn't sure what the traditional/classic/original was in the list of runners.  AKA - all y'all who have been on .3X or earlier you are using "blah runner"....  And your statement above that moving from the "traditional" scripted host runner that was the only choice for us early to the game.... should just work on 1.0 eluded me on this page and on the pages for each runner.   I "thought" maybe "traditional" is the "cake runner for .Net Framework".  I guess what threw me if that assumption is true, was the no .net core support in the table...

Hope this sheds some light on my confusion/misunderstanding.

**Reply by augustoproiete** [reply]
Interesting. I also saw this a couple of times the other way around where some devs wanted to use the .NET Tool runner to build .NET Framework 4.x projects and assumed it was not possible thinking that the Cake runner needed to match the same framework (.NET Framework or .NET Core) of the project that they are building.

**Reply by klabranche** [reply]
wait... are u saying .Net tool runner (dotnet cake) can compile Full Framework projects?  (I thought I tried that and it didn't work.... in another project....)

**Reply by pascalberger** [reply]
Yes, the blue annotation block mentions this above the [Supported platforms table](https://cakebuild.net/docs/running-builds/runners/#supported-platforms). Was this not obvious enough?

There actually hasn't changed anything regarding platform support or runners between 0.33 and 1.0. What has changed is that we improved documentation (which mostly was for Cake Runner for .NET Framework) to mention all available runners.

**Reply by augustoproiete** [reply]
Exactly... (couple of times++) :smile:.

Cake doesn't compile your code... It's just an orchestrator: It calls the tools that compile your code - and you decide which tools to call, so it's all about the tools/aliases that you use. If you call [MSBuild](https://cakebuild.net/dsl/msbuild/) with the version that you need, etc. it will do just that.

The .NET version of the Cake runner has nothing to do with the  .NET version of your projects.

**Reply by klabranche** [reply]
Oh boy... that block didn't even register with me (not your fault).  The supported platforms table really "sealed" the deal on what is supported in my mind.  Again, that's my bad.

**Reply by gitfool** [reply]
Yes, although you'll need to use the new [csproj](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj) format for projects.

**Reply by augustoproiete** [reply]
> Yes, although you'll need to use the new [csproj](https://docs.microsoft.com/en-us/dotnet/core/tools/csproj) format for projects.

Really? Why? That has not been my experience. I have a number of projects using the old .csproj format being built by Cake, using the dotnet tool.

**Reply by klabranche** [reply]
Ya... thinking that's why it didn't work.  The project was an original FF, then added a core 1 proj, that was upgraded to 2, etc....   this the csproj is still the old....

**Reply by klabranche** [reply]
I'll have to give this current project a try with dotnet cake instead at some point.  Don't have time atm. I already am using it on core only projects....

**Reply by gitfool** [reply]
>Really? Why? That has not been my experience. I have a number of projects using the old .csproj format being built by Cake, using the dotnet tool.

Maybe I'm misremembering since I switched to the new csproj format for all projects, old (full framework) and new (dotnet core), a long time ago.

**Reply by pascalberger** [reply]
It should really not make a difference if your buils runs on .NET Core or .NET Framework. In both cases it in the end just calls out to `msbuild` or `dotnet`. There's no hidden magic involved in here, only the Cake build script starting a process.

**Reply by klabranche** [reply]
Understood. :-)

**Reply by augustoproiete** [reply]
If it helps, [here is an example](https://github.com/augustoproiete/ookii-dialogs-winforms/tree/e23708ebf19214022302bd73ea4d9f0fafa46f44) of a Cake script that builds a .csproj in the old format using the dotnet cake tool.

**Reply by klabranche** [reply]
Not that you need my assurance but can confirm that cake tool runner is able to work with ff & core just fine, new and old csprojs... Just converted my project to use dotnet cake instead.  build.cake worked without any changes accept for one addin that isn't compatible with 1.0.... luckily was easy enough for me to get around that issue. :-)
